lib/src/signed_video_h26x_auth.c

```diff
@@ -115,6 +115,112 @@ decode_sei_data(signed_video_t *self, const uint8_t *payload, size_t payload_siz
+ * This function iterates through the NALU list, identifies the NALUs that belong to the current
+ * GOP, and marks them as used in the GOP hash. It then copies the hashes of these NALUs to the
+ * `nalu_hash_list` for later verification of the GOP hash. Additionally, it handles specific cases
+ * such as moving linked hash values for I-frames and ensures that SEI-related NALUs are
+ * appropriately handled.
+ */
+void
+prepare_for_verify_link_and_gop_hashes(signed_video_t *self, h26x_nalu_list_item_t *sei)
+{
+  // Ensure the `self` pointer is valid
+  assert(self);
+
+  // Initialize pointers and variables
+  h26x_nalu_list_t *nalu_list = self->nalu_list;
+  uint8_t *hash_list = &self->gop_info->nalu_hash_list[0];
+  int *list_idx = &self->gop_info->nalu_list_idx;
+  assert(nalu_list);
+  const size_t hash_size = self->verify_data->hash_size;
+  h26x_nalu_list_item_t *item = NULL;
+  uint8_t *hash_to_add = NULL;
+  bool found_next_gop = false;
+  bool found_item_after_sei = false;
+  uint8_t linked_hashes[hash_size * 2];
+  memset(linked_hashes, 0x00, sizeof(linked_hashes));
+
+  h26x_nalu_list_print(nalu_list);
+
+  // Start with the first item in the NALU list and reset the list index.
+  item = nalu_list->first_item;
+  *list_idx = 0;
+  remove_used_in_gop_hash(self->nalu_list);  // Clear previous `used_in_gop_hash` flags.
+
+  // Iterate through the NALU list until the end of the current GOP or SEI item is found.
+  while (item && !(found_next_gop || found_item_after_sei)) {
+    // Skip non-pending items
+    if (item->validation_status != 'P') {
+      item = item->next;
+      continue;
+    }
+
+    // Ensure that only non-missing NALUs (which have non-null pointers) are processed.
+    assert(item->nalu);
+
+    // Handle I-frame NALUs by managing linked hashes for previous GOP verification.
+    if (item->nalu->is_first_nalu_in_gop) {
+      memmove(&linked_hashes[0], &linked_hashes[hash_size], hash_size);
+      memcpy(&linked_hashes[hash_size], item->hash, hash_size);
+    }
+
+    // Track when the current item follows the SEI or marks the beginning of the next GOP.
+    found_item_after_sei = (item->prev == sei);
+    found_next_gop = (item->nalu->is_first_nalu_in_gop && !item->need_second_verification);
+
+    // Skip GOP SEI items as they do not contribute to the GOP hash.
+    if (item->nalu->is_gop_sei) {
+      item = item->next;
+      continue;
+    }
+
+    hash_to_add = item->need_second_verification ? item->second_hash : item->hash;
+    // Copy the hash to the list if there is enough space
+    if (*list_idx >= 0 && (*list_idx + hash_size <= self->gop_info->hash_list_size)) {
+      memcpy(&hash_list[*list_idx], hash_to_add, hash_size);
+      *list_idx += hash_size;
+      item->used_in_gop_hash = true;  // Mark the item as used in the GOP hash
+    }
+    // Move to the next item in the NALU list
+    item = item->next;
+  }
+
+  // Finally, mark the SEI item as used in the GOP hash
+  sei->used_in_gop_hash = true;
```

Comment:
But the SEI is no longer part of the GOP hash. Why do you need to mark it?
In fact, it may be marked as 'N' even if the signature was successfully verified, hence should be marked '.'